rust/hyperlane-base/src/server/server.rs

```diff
@@ -0,0 +1,113 @@
+use axum::{
+    http::{Response, StatusCode},
+    routing::get,
+    Router,
+};
+use hyper::Body;
+use prometheus::{Encoder, Registry};
+use std::{net::SocketAddr, sync::Arc};
+use tokio::task::JoinHandle;
+use tracing::warn;
+
+/// A server that serves agent-specific routes
+pub struct Server {
+    listen_port: u16,
+    registry: Registry,
+}
+
+impl Server {
+    /// Create a new server instance.
+    pub fn new(listen_port: u16, registry: Registry) -> Self {
+        Self {
+            listen_port,
+            registry,
+        }
+    }
+
+    /// Run an HTTP server serving agent-specific different routes
+    ///
+    /// routes:
+    ///   - metrics - serving OpenMetrics format reports on `/metrics`
+    ///     (this is compatible with Prometheus, which ought to be configured to scrape this endpoint)
+    pub fn run(self: Arc<Self>) -> JoinHandle<()> {
+        let port = self.listen_port;
+        tracing::info!(port, "starting prometheus server on 0.0.0.0");
+
+        let server_clone = self.clone();
+        tokio::spawn(async move {
+            let app = Router::new().route(
+                "/metrics",
+                get(move || {
```

Comment:
rather than have a closure here, let's move this logic to a standalone function like in the examples here: https://docs.rs/axum/latest/axum/#responses